frame/support/src/storage/bounded_vec.rs

```diff
@@ -19,61 +19,77 @@
 //! or a double map.
 
 use sp_std::prelude::*;
-use sp_std::{convert::TryFrom, marker::PhantomData};
+use sp_std::{convert::TryFrom, fmt, marker::PhantomData};
 use codec::{FullCodec, Encode, EncodeLike, Decode};
-use core::{ops::{Index, IndexMut}, slice::SliceIndex};
+use core::{
+	ops::{Deref, Index, IndexMut},
+	slice::SliceIndex,
+};
 use crate::{
 	traits::{Get, MaxEncodedLen},
 	storage::{generator, StorageDecodeLength, StorageValue, StorageMap, StorageDoubleMap},
 };
 
-/// Marker trait for types `T` that can be stored in storage as `BoundedVec<T, _>`.
-pub trait BoundedVecValue: FullCodec + Clone + sp_std::fmt::Debug {}
-impl<T: FullCodec + Clone + sp_std::fmt::Debug> BoundedVecValue for T {}
-
 /// A bounded vector.
 ///
 /// It has implementations for efficient append and length decoding, as with a normal `Vec<_>`, once
 /// put into storage as a raw value, map or double-map.
 ///
 /// As the name suggests, the length of the queue is always bounded. All internal operations ensure
 /// this bound is respected.
-#[derive(Encode, Decode, crate::DefaultNoBound, crate::CloneNoBound, crate::DebugNoBound)]
-pub struct BoundedVec<T: BoundedVecValue, S: Get<u32>>(Vec<T>, PhantomData<S>);
+#[derive(Encode, Decode)]
+pub struct BoundedVec<T, S>(Vec<T>, PhantomData<S>);
```

Comment:
What's the benefit of relaxing `S`? I can't see why and how that should ever be anything other than `Get<_>`? is it just to avoid writing it in a few places? I'd argue that's not worth it given you now need to split the `impl<T, S> BoundedVec<T, S>` into two chunks.

Reply:
I mean, yes, it's so that we don't have to write it down in as many places. Agree that `S` is never going to be anything other than `Get<_>`, but there are quite a few impls which just don't care about `S`, so there's not much point in enforcing the restriction there.

I suspect, but cannot prove, that more-general generics are slightly faster to compile: the compiler doesn't need to do as much work proving that the bounds are satisfied.

Reply:
In the meantime I looked a bit more in the std-library, and it indeed seems like the correct paradigm is to name as _few_ generics as possible, so you are indeed aligned with the idiomatic way of writing this kind of stuff. 

What I don't get and disagree with, is that it seems like this rule of thumb is stretched the extent that it even doesn't make sense anymore 😬

But anyway, thanks for the explanation. I am unfamiliar with some idioms.

Reply:
The point is, never bound a generic to a trait in the declaration if that is not required by the declaration. There are some exceptions of this rule, but here it is completely fine to drop it.